[Alternative] Move Parameter to Rust

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Based on the excellent work in #14207, this is PR implements a slightly alternative approach to porting the `ParameterExpression` to Rust, based on discussions with @doichanj, @kevinhartman and @mtreinish. The main differences are outlined below. We should compare both approaches and check which one is closer to what we'd like. Of course the efforts could also be merged.

### Details and comments

This PR does not supersede #14207, it is merely a slightly different approach and it's majority is based on #14207. The main differences are:
* Keep `ParameterExpression`, `Parameter` and `ParameterVectorElement` fully in Rust, without a Python-side class. This is done to reduce the code base and maintainer effort. This could also improve performance as we need less back and forth between the Rust and Python boundaries.
* Store the symbol's UUID on the `Symbol` in the symbolic expression engine itself, rather than keep a separate map of symbols to their UUID. This is done to couple the UUID and symbol closer instead of relying on a third object to make this connection.
* Implement pickling via QPY instead of string serialization -- we already have a serialization format which we could use. Maybe this allows to remove string-based operations in the future for safety?
* ... some other bits I probably forgot

Things that could make this PR better (maybe in this PR or a follow up)
- [x] Keep a separate `ParameterExpression` and `PyParameterExpression` for clearer Python split. Doing this would allow us to make the expression's `name_map` use `Symbol`s directly. Let's think about how we want to use this from C!
- [x] Make `Symbol` not a `pyclass` (probably easy to do, just didn't get to it yet)
- [x] Should we remove the index from `Symbol` and have an `IndexedSymbol`, like in #14207? Answer: no, it won't reduce memory footprint, so we don't bother now.
* ~~try to remove string parsing~~ this will be a follow up
* ... ~~some other bits I surely did forget and am unaware of~~

Co-authored-by: "U-AzureAD\\JUNDOI" <doichan@jp.ibm.com>

Cryoris committed:
Split py and py-free methods

Cryoris committed:
playing with Symbol as Parameter

-- 606 failing tests

Cryoris committed:
fix errors

Cryoris committed:
about to start the mess

Cryoris committed:
suite passing

Cryoris committed:
PVE in Rust

Cryoris committed:
start assign

Cryoris committed:
add assign

Cryoris committed:
dealing with unknown params and conflicts

Cryoris committed:
an attempt at vector fixing

Cryoris committed:
fix more tests

- name conflicts
- hash(2.3) == hash(ParamExpr(2.3))
- some error messages

Cryoris committed:
implement name_map as track record

This allows keeping track of which parameters are still in the expression, even though the expression is optimized. E.g. x*0 = 0, but x is still valid to bind.

Cryoris committed:
The string debacle

- don't use UUID in to_string, only in string_id (using strings for comparisons should probably be avoided in future)
- clippy
- relax Python is-checks to eq-checks

Cryoris committed:
some housekeeping

Cryoris committed:
implement _qpy_replay

Cryoris committed:
implement sympify

Cryoris committed:
fix copy/deepcopy

Cryoris committed:
Implement pickle via QPY and Parameter.assign to return Parameter

Cryoris committed:
fix template matching

this is rather a temporary solution, maybe we ought to disable string parsing in favor or a proper SymPy->parameter expression conversion

Cryoris committed:
fix rhs - lhs mixup

Cryoris committed:
fix more tests on binding {x: 0*y}

Cryoris committed:
fix last failing tests (see numeric(strict=False))

Cryoris committed:
clippy

Cryoris committed:
a lot of docs

Cryoris committed:
rename PyParameterExpression to ParameterExpression

Cryoris committed:
Merge branch 'main' into rust-param-pyparam

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2022 @@
+    #[error("Cannot bind Parameter {0:?} not present in expression")]
+    UnknownParameter(Symbol),
+    #[error("Cannot bind following parameters not present in expression: {0:?}")]
+    UnknownParameters(HashSet<Symbol>),
```
These should be combined I'd think, since otherwise callers have to handle both errors.

Cryoris committed:
Thou shan't fake hyperrefs

kevinhartman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2332,7 +2332,13 @@ impl CircuitData {
          -> PyResult<Param> {
+            // let as_str = expr.call_method0("__str__")?;
+            // let params = expr.getattr(parameters_attr)?;
+            // println!("expr {as_str:?} {params:?}");
```
```suggestion
```

kevinhartman commented on crates/circuit/src/circuit_data.rs:
```diff
@@ -2332,7 +2332,13 @@ impl CircuitData {
             let new_expr = expr.call_method1(assign_attr, (param_ob, value.into_py_any(py)?))?;
+            // let as_str = new_expr.call_method0("__str__")?;
+            // let params = new_expr.getattr(parameters_attr)?;
+            // println!("new_expr {as_str:?} {params:?}");
```
```suggestion
```

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2022 @@
+    }
+}
+
+/// Attempt to extract a [PyParameterExpression] from a bound [PyAny].
```
```suggestion
/// Attempt to extract a [ParameterExpression] from a bound [PyAny].
```

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2022 @@
+    // A map keeping track of all symbols, with their name. This map *must* be kept
+    // up to date upon any operation performed on the expression.
+    // TODO it would be nicer to just store [Symbol]s as values, which can maybe be done
+    // with a secondary PyParameterExpression storing the [PyParameter] in a map.
```
This comment seems to imply that `ParameterExpression`s are mutable. I guess they are technically because these fields are public, but would it be better if they weren't? It looks like in most cases any operation on a `ParameterExpression` results in a brand new instance.

Cryoris replied:
They are immutable -- this was a comment targeting devs 🙂 With only the double backslash this shouldn't show up in any docs, right?

Cryoris replied:
It's a good idea to make these attributes private. I think some attributes I just marked public because I needed them at some points where they weren't accessible. I'll go through to check that they are correctly private 👍🏻

Cryoris replied:
Done -- the only one not private is `PyParameterExpression.inner`, which is used in the parameter table right now. This could also be changed if we want, but probably this will be resolved in the follow up PR anyways.

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2022 @@
+        }
+    }
+
+    /// Substitute symbols with [PyParameterExpression]s.
```
```suggestion
    /// Substitute symbols with [ParameterExpression]s.
```

Probably should do a replace-all for this.

doichanj commented on crates/circuit/src/parameter/symbol_parser.rs:
```diff
@@ -74,9 +76,9 @@ fn parse_symbol(s: &str) -> IResult<&str, SymbolExpr, VerboseError<&str>> {
                     // add indexing in Symbol struct
                     let s = format!("{v}[{i}]");
-                    Ok(SymbolExpr::Symbol(Arc::new(s)))
+                    Ok(SymbolExpr::Symbol(Symbol::new(&s, None, None)))
```
Here we parse index of vector element, so Symbol can be initialized as
```suggestion
                    Ok(SymbolExpr::Symbol(Symbol::new(v, None, Some(i))))
```

Cryoris replied:
Good catch! Done 👍🏻

Cryoris committed:
proper split: (Py)ParameterExpression

kevinhartman reviewed (commented):
Adding just a couple of comments before you push the `PyParameterExpression` split stuff.

Cryoris committed:
rm pyclass from Symbol

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+    // The symbolic expression.
+    pub expr: SymbolExpr, // TODO make these attrs private?
+    // A map keeping track of all symbols, with their name. This map *must* be kept
+    // up to date upon any operation performed on the expression.
```
The reason this comment bugs me is that it makes it seem like as a developer, I should expect to mutate a `ParameterExpression` while performing operations that involve it. But it sounds like from your earlier reply that this would only ever be set once during construction. Maybe tweak it to say "This map *must* have entries for all symbols used by the expression."

Cryoris replied:
Ah yeah that's more precise -- I'll update it.

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+
+impl Hash for ParameterExpression {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.expr.to_string().hash(state);
```
Just to be sure, it's not possible for two different expressions from a tree-perspective to have the same string representation, correct?

Cryoris replied:
Hm I think this should be `string_id()` and not `to_string()` since the current version will not take into account UUIDs and could give the same hash for expressions that use the same parameter name but different instances. Other than that, the string representation should be unique as I understand it, since it's build by traversing the tree. I'll update to `string_id()`.

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+            } else {
+                match self.expr.eval(true) {
+                    Some(e) => e.to_string(),
+                    None => self.expr.optimize().to_string(),
```
Why should we evaluate or optimize the expression? Isn't it better to just display it exactly as it is? Imagine if we compare two expressions for equality and find that they aren't equal, but then when we print, they evaluate to the same thing.

Cryoris replied:
Good question, this was here before.. let me try removing this and see what happens

Cryoris replied:
Removed 👍🏻

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+    ///
+    /// Caution: The caller **guarantees** that ``name_map`` is consistent with ``expr``.
+    /// If uncertain, call [Self::from_symbol_expr], which automatically builds the correct name map.
+    pub fn new(expr: &SymbolExpr, name_map: &HashMap<String, Symbol>) -> Self {
```
What is the use-case for creating a custom mapping, rather than reading it from the `SymbolExpr`? 

In any case, this function should at least take both of its inputs by value and force the caller to do the clone. This is better practice in general since it is honest about the ownership needs of the call.

Cryoris replied:
We could read the `name_map` from the `SymbolExpr`, but it's not for free and in some cases we already know the `name_map`: e.g. when we create an expression from a value (in `PyParameterExpression::extract_coerce`) or in `PyParameterExpression::py_new`. If we know the `name_map` already we don't need to parse the `SymbolExpr` again to construct it. That being said I didn't benchmark how much of a difference that would make, probably not much 🙂 

I'll update the cloning 👍🏻

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+    }
+
+    /// Construct from a [Symbol].
+    pub fn from_symbol(symbol: &Symbol) -> Self {
```
```suggestion
    pub fn from_symbol(symbol: Symbol) -> Self {
```

Same comment

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+            expr: expr.clone(),
+            name_map: expr.name_map().clone(),
+        }
+    }
```
Do like this 🙂 

```suggestion
    pub fn from_symbol_expr(expr: SymbolExpr) -> Self {
        let name_map = expr.name_map();
        Self {
            expr,
            name_map,
        }
    }
```

Otherwise you've got extra clones you certainly don't want. In your current version, I think the `.clone()` on `name_map()` is actually fully unnecessary anyhow!

Cryoris replied:
ah that's much better, thanks !

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+                OpCode::EXP => lhs.exp(),
+                OpCode::SIGN => lhs.sign(),
+                OpCode::GRAD | OpCode::SUBSTITUTE => {
+                    unreachable!("GRAD and SUBSTITUTE are not supported.")
```
I would use `panic!`. I think we should reserve `unreachable!` for cases where it's truly supposed to be impossible for program control to reach it. I can craft a qpy replay that'll trigger this, so it's not unreachable 🙂 .

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+        // once we're done, just return the last element in the stack
+        Ok(stack
+            .pop()
+            .expect("OPReplay was empty, no expression could be built"))
```
This would imply that the replay was invalid, right? Maybe a better message would be "invalid qpy replay encountered during deserialization"

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+    pub fn add(&self, rhs: &ParameterExpression) -> Result<Self, ParameterError> {
+        let name_map = self.update_name_map(rhs)?;
+        Ok(Self {
+            expr: &self.expr + &rhs.expr,
```
Just a thought, but right now this line will end up calling the `Add` impl on `&SymbolExpr`, which ends up cloning both `lhs` and `rhs`. I wonder if it'd be more efficient if we made `ParameterExpression::expr` an `Arc<SymbolExpr>` after all and changed `Add` to be implemented for `Arc<SymbolExpr>`.

Perhaps a rabbit hole to nowhere. It just feels like we clone things a lot more than I'd expect in all of this (an existing issue, not introduced here).

Cryoris replied:
Good point, perhaps something to keep as follow up? I can open an issue to keep track for such performance issues 🙂

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+
+    /// Clone the expression.
+    pub fn copy(&self) -> Self {
+        // TODO: Is this necessary? Doesn't ParameterExpression.clone() implement the same thing?
```
Yes it would. We derive `Clone` and this is what that would look like. Go ahead and remove this method IMO.

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+    /// Args:
+    ///     - map: A hashmap with [Symbol] keys and [ParameterExpression]s to replace these
+    ///         symbols with.
+    ///     - allow_unknown_parameters: If ``false``, returns an error if any symbol in the
```
Just one backtick in Rust docstring land, I believe.

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+                .collect();
+            if !difference.is_empty() {
+                return Err(ParameterError::UnknownParameters(difference));
+            }
```
Can you defer the clone until we know we're in an error path?

```suggestion
            let difference: HashSet<&Symbol> = to_replace
                .difference(&existing)
                .collect();
            if !difference.is_empty() {
                return Err(ParameterError::UnknownParameters(difference.cloned()));
            }
```

(untested)

Cryoris replied:
That's a good point -- I changed it to the following trying also to avoid collecting the `difference` iterator. What do you think?

```rust
            let mut difference = to_replace.difference(&existing).peekable();

            if difference.peek().is_some() {
                let different_symbols = difference.map(|s| (**s).clone()).collect();
                return Err(ParameterError::UnknownParameters(different_symbols));
            }
```
I'm not very fond of the `**s` bit, but I guess this is the way?

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+        allow_unknown_parameters: bool,
+    ) -> Result<Self, ParameterError> {
+        // The set of symbols we will bind. Used twice, hence pre-computed here.
+        let bind_symbols: HashSet<Symbol> = map.keys().cloned().collect();
```
Can we use `HashSet<&Symbol>` here?

Cryoris replied:
Yes we can (and do now!)

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+    ///
+    /// * `Ok(HashMap<String, Symbol>)` - The merged name map.
+    /// * `Err(ParameterError)` - An error if there was a name conflict.
+    fn update_name_map(&self, other: &Self) -> Result<HashMap<String, Symbol>, ParameterError> {
```
```suggestion
    fn merged_name_map(&self, other: &Self) -> Result<HashMap<String, Symbol>, ParameterError> {
```

Otherwise this is confusing unless you click in to see that it accepts `&self` not `&mut self`.

Cryoris replied:
also the doc says "merge name maps" lol

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+/// for the parameters inside the expression.
+#[pyclass(
+    subclass,
+    sequence,
```
Is this really a sequence type?

Cryoris replied:
Hm I don't know why this was here, the original Python class doesn't seem to be a sequence type and the tests also pass without this -- removed!

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+)]
+#[derive(Clone, Debug)]
+pub struct PyParameterExpression {
+    pub inner: Arc<RwLock<ParameterExpression>>,
```
Did you end up trying to make this just a `ParameterExpression`?

Cryoris replied:
Yep, but didn't push yet -- sorry!

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+///
+/// This is backed by Qiskit's symbolic expression engine and a cache
+/// for the parameters inside the expression.
+#[pyclass(
```
I think you can just add `hash` to this list if you want to skip having `__hash__`.

Cryoris replied:
I added a custom `__hash__` because we have some tests relying on the fact that `hash(<some number>) == hash(ParameterExpression(<some number>))`. Not sure we want this, but to keep the existing behavior I couldn't use the default hash implementation 🤔

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+pub struct PyParameterExpression {
+    pub inner: Arc<RwLock<ParameterExpression>>,
+    // in contrast to ParameterExpression::name_map, this stores [PyParameter]s as value
+    pub name_map: HashMap<String, PyParameter>,
```
We already talked about this offline, but it feels like it's probably better not to have this mapping if we already have one inside `ParameterExpression`.

Cryoris replied:
Removed 👍🏻

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+            .name_map
+            .iter()
+            .map(|(name, symbol)| (name.clone(), PyParameter::new(symbol)))
+            .collect();
```
```suggestion
        let name_map = value
            .name_map
            .into_iter()
            .map(|(name, symbol)| (name, PyParameter::new(symbol)))
            .collect();
```

maybe?

Cryoris replied:
Sadly not because the next line consumes `value` 😄

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+    fn extract_coerce(ob: &Bound<'_, PyAny>) -> PyResult<Self> {
+        if let Ok(i) = ob.extract::<i64>() {
+            Ok(
+                ParameterExpression::new(&SymbolExpr::Value(Value::from(i)), &HashMap::new())
```
Yes, this callsite definitely looks like it wants to take its arguments by value!

Cryoris replied:
yep I realized this when changing the input type to `new` 😛

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,2162 @@
+
+    /// Check if the expression corresponds to a plain symbol.
+    ///
+    /// TODO can we delete this? Not part of public interface before.
```
If we don't, then it will be. Just making a note here to delete or prefix with `_`.

Cryoris replied:
We rely on this this in `sympify` and `qpy`. I would say if we rely on this then we might as well expose it? But I don't have a strong opinion here

kevinhartman reviewed (commented):
Not a complete review yet, but submitting what I have so far. I'll plan to finish up on Monday 🙂.

Cryoris committed:
TODO's and no more RwLock

- rm Arc<RwLock<..>> for inner, since expressions are immutable and not shared, this is not needed
- address a bunch of todos, that include
  - rm rsub/rdiv/rpow
  - introduce try_to_value and _to_symbol
  - make internal data private

Cryoris committed:
First round of review comments

mainly removing some unnecessary clones, and fixing Hash for ParameterExpression

Cryoris committed:
rm `name_map` from PyParameterExpression

--- and just reconstruct the Py symbols from ParameterExpression.name_map

Cryoris commented:
I ran some preliminary benchmarks comparing Qiskit v2.1.1 vs this and it seems this PR speeds things up a bit:
```
-- Construct + bind efficient_su(num_qubits=n, reps=n)
(our PR)
--- n=100 took 1.002s
--- n=200 took 4.394s
--- n=500 took 30.157s
(2.1.1)
--- 100 took 1.180s
--- 200 took 5.945s
--- 500 took 41.248s

-- Compute sum_{i=1}^n (x_i - i/100) ** 2
(our PR)
--- 100 took 0.008s
--- 200 took 0.026s
--- 500 took 0.104s
(2.1.1)
--- 100 took 0.133s
--- 200 took 0.070s
--- 500 took 0.397s
```
Note that the first benchmarks measures time to _construct_ the circuit too, so the binding speedup is hidden a bit.

Cryoris committed:
Fix vector element parsing

Co-authored-by: "U-AzureAD\JUNDOI" doichan@jp.ibm.com

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -172,32 +147,71 @@ impl ParameterExpression {
+    /// Try casting to a [Symbol].
+    ///
+    /// This only succeeds if the underlying expression is, in fact, only a symbol.
+    pub fn try_to_symbol(&self) -> Result<Symbol, ParameterError> {
```
My first thought was that we tend to just return an `Option` for these Rust "downcast"-like functions, but I see that with `try_to_value` below it makes a lot of sense to (as you have) return an error e.g. for unbound parameters. So no change requested to keep consistent with that method.

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -229,11 +243,11 @@ impl ParameterExpression {
                 OpCode::ADD => lhs.add(&rhs.unwrap())?,
                 OpCode::MUL => lhs.mul(&rhs.unwrap())?,
                 OpCode::SUB => lhs.sub(&rhs.unwrap())?,
-                OpCode::RSUB => lhs.rsub(&rhs.unwrap())?,
```
Neat, less code!

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -648,29 +617,23 @@ impl ParameterExpression {
+    pub inner: ParameterExpression,
 }
 
 impl Hash for PyParameterExpression {
```
Do we need an impl of `Hash` and `PartialEq` on the Python wrapper types? I wonder if it's better not to have them to encourage folks to use the inner types directly in Rust space.

Cryoris replied:
Hm good point this might also be legacy, I'll try removing it

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -970,15 +886,17 @@ impl PyParameterExpression {
+        let d_expr = self
+            .inner
+            .derivative(&symbol)
+            .map_err(PyRuntimeError::new_err)?;
```
I was surprised to find that `SymbolExpr::derivative` (which is called by `ParameterExpression::derivative` called here) returns a `Result` with a `String` error type. Maybe something to address in the future.

Cryoris replied:
Yeah there's not really a SymbolError type. Did you expect one?

kevinhartman replied:
I suppose I did, probably just because `ParameterExpression` has an error type, and `SymbolExpr` does a lot of the lifting for it.

Cryoris replied:
I did a compromise solution: I left `SymbolExpr` have the `String` error type, but I updated `ParameterExpression::derivative` to return a `ParameterError`, consistent with the rest 🙂

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -1016,15 +934,11 @@ impl PyParameterExpression {
         // reduce the map to a HashMap<Symbol, ParameterExpression>
         let map = parameter_map
             .iter()
```
I think `into_iter()` works here, and then you can skip the clones later.

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,1984 @@
+
+    /// Sine of the expression.
+    #[pyo3(name = "sin")]
+    pub fn py_sin(&self) -> Self {
```
I wonder if there'd be any benefit to marking these kinds of wrapper methods as `#[inline]`. I think that PyO3 generates those "trampoline" functions under the hood that then call the actual method in question, so maybe making these `#[inline]` would mean the call to `self.inner.sin()` always gets embedded into the trampoline function's code directly.

Cryoris replied:
Changed to `inline` 👍🏻

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,1984 @@
+    ) -> PyResult<Self> {
+        // reduce the map to a HashMap<Symbol, Value>
+        let map = parameter_values
+            .iter()
```
`into_iter` to avoid later clone, I think

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,1984 @@
+    /// For backward compatibility only. This should not be used and we ought to update those
+    /// usages!
+    #[getter]
+    pub fn _vector(&self) -> PyObject {
```
The idea is to deprecate the public getter `vector` above, correct?

Cryoris replied:
We'd only simplify the model if we fully removed the `.vector` attribute, not just deprecate the getters. I'll open a discussion/issue on this later 🙂

kevinhartman commented on crates/circuit/src/parameter/parameter_expression.rs:
```diff
@@ -0,0 +1,1984 @@
+
+/// Replace [Symbol]s in a [SymbolExpr] according to the name map. This
+/// is used to reconstruct a parameter expression from a string.
+fn replace_symbol(symbol_expr: &SymbolExpr, name_map: &HashMap<String, Symbol>) -> SymbolExpr {
```
Should this be inside `symbol_expr.rs` instead?

kevinhartman commented on crates/circuit/src/parameter/symbol_expr.rs:
```diff
@@ -10,26 +10,97 @@
+pub struct Symbol {
+    name: String,           // the name of the symbol
+    pub uuid: Uuid,         // the unique identifier
+    pub index: Option<u32>, // an optional index, if part of a vector
```
For others, we discussed offline if it made sense to have a separate `IndexSymbol` type with the goal of keeping `index` out of `Symbol` for the majority case when we aren't representing an indexed symbol. The thinking was that this likely wouldn't reduce memory use since `Symbol` when used as a parameter would exist within `SymbolExpr`, which would be at least as big as any `IndexSymbol` anyway.

kevinhartman commented on crates/circuit/src/parameter/symbol_expr.rs:
```diff
@@ -10,26 +10,97 @@
+    name: String,           // the name of the symbol
+    pub uuid: Uuid,         // the unique identifier
+    pub index: Option<u32>, // an optional index, if part of a vector
+    #[allow(dead_code)]
```
Why is this dead code again?

Cryoris replied:
huh that's probably legacy -- I think this can be removed

kevinhartman commented on crates/circuit/src/parameter/symbol_expr.rs:
```diff
@@ -10,26 +10,97 @@
+        index: Option<u32>,
+        vector: Option<PyObject>,
+    ) -> PyResult<Self> {
+        if index.is_some() && vector.is_none() || index.is_none() && vector.is_none() {
```
Looks like a bug: `index.is_none() && vector.is_none()`

Is there a way we could add a test to expose this?

Separately, how about:

```suggestion
        if index.is_some() != vector.is_some() {
```

Cryoris replied:
Changed to your suggestion 👍🏻 From Python I don't think we can test this since this is only ever called in `PyParameterExpression.__setstate__` and we explicitly set index and vector to `Some(..)` 🤔

kevinhartman commented on crates/circuit/src/parameter/symbol_expr.rs:
```diff
@@ -710,36 +622,32 @@ impl SymbolExpr {
+            SymbolExpr::Symbol(e) => {
+                let mut set = HashSet::<Symbol>::new();
+                set.insert(e.clone()); // TODO can this be done more concisely?
+                set
```
```suggestion
                HashSet::from_iter([e.clone()])
```

Cryoris replied:
heh I knew there was a better way -- thanks!

kevinhartman commented on crates/circuit/src/parameter/symbol_expr.rs:
```diff
@@ -997,6 +904,10 @@ impl SymbolExpr {
     }
 
+    pub fn string_id(&self) -> String {
+        self.display(true)
```
This _looks_ sus lol. 

I think there's a risk that someone might change the implementation of `display` in the future, not realizing that it is critical to the formulation of this ID. What's the purpose of this wrapper function anyhow? Can you just rename `display` to `string_id`? 

...but also, this isn't really an ID, it's more of a representation, right? Maybe get rid of this method and just have a `SymbolExpr::string_repr`.

Cryoris replied:
I introduced this since the code was relying on `to_string()` for some equality comparisons, which missed the parameters UUID. To make this a little safer (maybe we ought to rework this later properly) I added the UUID for internal representation, though we don't want to print it for standard displaying --> hence the `fn display(with_uuid: bool)`. 

I can change this to `string_repr(with_uuid)` as you suggested 👍🏻

kevinhartman commented on crates/circuit/src/parameter/symbol_expr.rs:
```diff
@@ -2562,6 +2473,180 @@ impl SymbolExpr {
         }
     }
+
+    fn display(&self, with_uuid: bool) -> String {
```
Commenting again here. Please rename this to something else lol

kevinhartman commented on qiskit/circuit/rust_parameter_expression.py:
```diff
@@ -0,0 +1,38 @@
+# This code is part of Qiskit.
```
Is this file needed?

Cryoris replied:
Lol sorry I didn't realize this was still in the PR -- no it's not needed

kevinhartman commented on qiskit/qpy/binary_io/value.py:
```diff
@@ -26,30 +27,43 @@
 
 def _write_parameter_vec(file_obj, obj):
-    name_bytes = obj._vector._name.encode(common.ENCODE)
+    name_bytes = obj.vector._name.encode(common.ENCODE)
```
Can you remind me what the purpose of this change is? Aren't we getting rid of `vector` someday?

Cryoris replied:
I changed this since `.vector` exists as public attribute so we should prefer that over the private method. If we remove it in the future we'll have to deprecate it and remove it properly anyways, so we might as well use the public attribute in the meanwhile 🙂

kevinhartman commented on qiskit/qpy/binary_io/value.py:
```diff
@@ -170,15 +184,18 @@ def _write_parameter_expression(file_obj, obj, use_symengine, *, version):
-    symbol_table_len = len(obj._parameter_symbols)
+
+    parameters = obj.parameters
+    fake_symbols_dict = dict(zip(parameters, parameters))
```
It's not really "fake" is it? Heh. I think that name is pretty confusing, since the original `_parameter_symbols` is gone (...so anyone reading this will wonder what makes it "fake")

Cryoris replied:
Oh that was meant to be temporary, I'll rework this 👍🏻

kevinhartman commented on test/python/circuit/test_circuit_load_from_qpy.py:
```diff
@@ -740,6 +740,9 @@ def test_qaoa(self):
 
+        # new_params = [Parameter(f"{i}") for i in range(qaoa.num_parameters)]
+        # qaoa.assign_parameters(new_params, inplace=True)
+
```
Delete?

kevinhartman commented on test/python/circuit/test_parameter_expression.py:
```diff
@@ -477,7 +478,7 @@ def test_sympify_all_ops(self):
         d = Parameter("d")
 
-        expression = (a + b.sin() / 4) * c**2
+        expression = (a + b.sin() * 0.25) * c**2
```
Any commentary on why needed to be changed?

Cryoris replied:
Our symbolic engine turns `x/4` into `x*0.25`, so I had to change this for the tests to pass. I think Doi-san's follow up on `Rational`s would fix this, but I also don't think it's very important that we string-match `x/4` as long as the expression is mathematically correct.

kevinhartman commented on test/python/circuit/test_parameters.py:
```diff
@@ -1597,7 +1599,8 @@ def test_raise_if_subbing_in_parameter_name_conflict(self):
             expr.subs({x: y_second})
 
-    @data(2, 1.3, 0, -1, -1.0, numpy.pi, 1j)
+    # @data(2, 1.3, 0, -1, -1.0, numpy.pi, 1j)
```
Should you uncomment again?

kevinhartman commented on test/python/circuit/test_parameters.py:
```diff
@@ -2190,12 +2192,14 @@ def test_parameter_equal_to_identical_expression(self):
-        phi = Parameter("phi")
-        cos_phi = numpy.cos(phi)
-        self.assertEqual(phi._parameter_symbols, cos_phi._parameter_symbols)
+    # Julien: this test was using the internal ._name_map property -- rewrite using
```
Reminder for you to do this :)

kevinhartman reviewed (commented):
This completes my first pass of review. Thanks to both you and @doichanj for all of this work.

Overall, this looks good to me. The one high-level question I've got concerns the split between `PyParameterExpression`, `ParameterExpression`, and `SymbolExpr`. Am I correct in my understanding that the only reason to have a `ParameterExpression` is so we have some place to maintain a `name_map`? If not for this, I would think we could get away with just having `PyParameterExpression` for Python and `SymbolExpr` for Rust.

Cryoris commented:
> The one high-level question I've got concerns the split between PyParameterExpression, ParameterExpression, and SymbolExpr. Am I correct in my understanding that the only reason to have a ParameterExpression is so we have some place to maintain a name_map? If not for this, I would think we could get away with just having PyParameterExpression for Python and SymbolExpr for Rust.

Yeah I also thought about this (in fact it's what I had initially). There's 2 reasons I added the `name_map`:

1. If we only keep `SymbolExpr`, which is an enum, we can't store any cached states though, right? Meaning that if we repeatedly query the parameters we would have to traverse the tree each time. Since we do this quite a lot, e.g. for assigning parameters or checking for parameter conflicts, I thought it would be better to have the cache. 
2. The current symbolic expressions optimize things like `0*x` to `0` and lose the `Symbol`. However, if we do `qc.rx(0 * x, 0)` then Qiskit expects that I can still bind `x`. We need to keep track of this symbol somewhere, which is the `name_map`.

That being said -- the `ParameterTable` already does some caching for parameter assignment, and I didn't benchmark what happens when we remove that cache. I can try and see what happens, if we want 🙂 We might be able to use the same cache to check for the existence of `x` in `0*x`-type expressions.

An advantage of not having `ParameterExpression` would of course be one layer less, which is nicer to maintain.

Cryoris committed:
finish 1st review round

mainly rm unused code and dangling comments

Cryoris committed:
left over review comments

- dangling commented code
- error for derivative

Cryoris committed:
rename SymbolExpr::display --> repr

Co-authored-by: Jun Doi <doichan@jp.ibm.com>

kevinhartman commented on crates/circuit/src/parameter/symbol_expr.rs:
```diff
@@ -905,7 +900,7 @@ impl SymbolExpr {
     }
 
     pub fn string_id(&self) -> String {
```
I'm still not a fan of calling this an "id" since it sounds like what gets returned is a simple ID of a constant length.

kevinhartman reviewed (approved):
This looks good to me now. I have one minor comment, but I don't think we need to hold this up.

For posterity, we talked offline about the idea of eliminating `ParameterExpression` (and just keeping `SymbolExpr`) and our conclusion was that `SymbolExpr` is more of a Qiskit-independent symbol layer, and `ParameterExpression` is the piece that wraps it with our extra domain-specific logic. I suggested that we might want to move `SymbolExpr` to its own crate to avoid future coupling, but we've decided to defer that for a potential follow-up.
